src/plonk/circuit.rs

```diff
@@ -679,6 +695,46 @@ impl<F: Field> Expression<F> {
     pub fn square(self) -> Self {
         self.clone() * self
     }
+
+    /// Returns whether or not this expression contains a `Selector`.
+    fn contains_simple_selector(&self) -> bool {
```

Comment:
Potential performance issue: since this is called on each subexpression when adding, subtracting or multiplying expressions, the cost of constructing an expression is now quadratic in the number of nodes in its AST.

Reply:
Good point. I think if this gets out of hand we can address it, but note that none of these operations with `Expression` happen outside keygen (or tests) in practice.